contracts/two_prisoners_dilemma/lib.rs

```diff
@@ -0,0 +1,527 @@
+        vote_count: u32,
+
+        /// Registered player accounts
+        players: [AccountId; NUMBER_OF_PLAYERS],
+
+        /// Claimed players rewards
+        claimed_rewards: Mapping<AccountId, ()>,
+    }
+
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        /// sender is not registered player, either when trying to cooperate or defect
+        CallerIsNotPlayer,
+
+        /// player already chosen, ie cooperated or defected
+        PlayerAlreadyChosen,
+
+        /// not all players made their choices
+        NotAllPlayersMadeTheirChoices,
+
+        /// contract's funds are insufficient to pay out reward to a player
+        InsufficientContractFunds,
+
+        /// requested transfer failed, this can be the case if the contract does not
+        /// have sufficient free funds or if the transfer would have brought the
+        /// contract's balance below minimum balance
+        TransferFailed,
+
+        /// a reward has been already claimed
+        RewardAlreadyClaimed,
+    }
+
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    impl TwoPrisonersDilemma {
+        /// Creates new instance of Prisoner's Dilemma SC with below parameters:
+        /// # Arguments
+        ///
+        /// * `players` - list of exactly 2 accounts what will play the game
+        ///
+        /// Game has following rewards configured (in tokens):
+        /// * `bad_luck_payoff`: 0
+        /// * `defect_payoff`: 1
+        /// * `cooperation_payoff`: 2,
+        /// * `temptation_payoff`: 3.
+        #[ink(constructor)]
+        pub fn new(players: [AccountId; 2]) -> Self {
+            initialize_contract(|contract: &mut Self| {
+                contract.players = players.clone();
+                for player in players {
+                    contract.choices.insert(&player, &Choice::Waiting);
+                }
+
+                contract.bad_luck_payoff = 0;
+                contract.defect_payoff = TOKEN;
+                contract.cooperation_payoff = 2 * TOKEN;
+                contract.temptation_payoff = 3 * TOKEN;
+
+                contract.vote_count = 0;
+            })
+        }
+
+        fn get_callers_current_choice(&self, caller: &AccountId) -> Result<Choice> {
+            let current_choice = self.choices.get(caller);
+            if current_choice.is_none() {
+                return Err(Error::CallerIsNotPlayer);
+            }
+            Ok(current_choice.unwrap())
+        }
+
+        fn get_other_player_choice(&self, this_player: &AccountId) -> Choice {
+            let other_player_index: usize = match self.players[0] == *this_player {
+                true => 1,
+                false => 0,
+            };
+            return self.choices.get(self.players[other_player_index]).unwrap();
+        }
+
+        fn make_choice(&mut self, choice: &Choice) -> Result<()> {
```

Comment:
you should conceal the choice, for example by submitting a keccak256(Choice, secret) and storing that in a AccountId => Hash mapping. Than you should have a reveal_choice with arguments Choice, secret that check whether it hashes to the same value.

There might be people that don't reveal in time: you could alter the make_choice to take a said hash and a plaintext FakeChoice, and using that when comes the time to reveal, something like that

Reply:
Right, as per blinded auction https://docs.soliditylang.org/en/v0.8.16/solidity-by-example.html#id2 right? What I don't understand is how one computes this a keccak256(Choice, secret) on the side. 

Btw do you know what's the equivalent of Solidity's `payable` in `ink!`'s world? E.g with below call user somehow transfers tokens to contract?
```
function bid(bytes32 blindedBid)
        external
        payable
        onlyBefore(biddingEnd)
    {
        bids[msg.sender].push(Bid({
            blindedBid: blindedBid,
            deposit: msg.value
        }));
    }
```

Reply:
You have `#[ink(message, payable)]`.

Reply:
> What I don't understand is how one computes this a keccak256(Choice, secret) on the side.

If it was a game with a frontend then you could hash the input on user's side (in the browser) and never leak it oustide.

Reply:
Turns out it is enough to paste sth llike this in https://test.azero.dev/#/js

>console.log(hashing.keccakAsHex("secret"));
console.log(hashing.keccakAsU8a("secret"));

Ouput

>0x65462b0520ef7d3df61b9992ed3bea0c56ead753be7c8b3614e0ce01e4cac41b
101,70,43,5,32,239,125,61,246,27,153,146,237,59,234,12,86,234,215,83,190,124,139,54,20,224,206,1,228,202,196,27

Reply:
Oh yes, I thought you were asking how to automate it and take the burden off of the user.